pkg/compute/bidder.go

```diff
@@ -104,21 +124,28 @@ func (b Bidder) ReturnBidResult(ctx context.Context, execution store.Execution,
 func (b Bidder) doBidding(
 	ctx context.Context,
 	bidStrategyRequest bidstrategy.BidStrategyRequest,
-	jobRequirements model.ResourceUsageData,
-) (*bidstrategy.BidStrategyResponse, error) {
-	// Check bidding strategies before having to calculate resource usage
-	bidStrategyResponse, err := b.strategy.ShouldBid(ctx, bidStrategyRequest)
+	calculator capacity.UsageCalculator,
+) (*bidstrategy.BidStrategyResponse, *model.ResourceUsageData, error) {
+
+	// Check semantic bidding strategies before having to calculate resource usage
+	bidStrategyResponse, err := b.semanticStrategy.ShouldBid(ctx, bidStrategyRequest)
 	if err != nil {
-		return nil, fmt.Errorf("error asking bidding strategy if we should bid: %w", err)
+		return nil, nil, fmt.Errorf("error asking bidding strategy if we should bid: %w", err)
 	}
 
+	// if the request is semantically biddable check resource-based bidding.
+	var resourceUsage model.ResourceUsageData
 	if bidStrategyResponse.ShouldBid {
```

Comment:
If the semantic strategy responds with `ShouldBid: false, ShouldWait: true`, which is what will happen with using an async HTTP call, this will cause the resourceUsage config to not be evaluated, so the execution will be created with a nil resource usage. (The system effectively ignores the value of ShouldBid if ShouldWait is true, at the moment. So a tri-state value might have been a better choice.)

We then never go back to compute it or even run the resource-based strategies, because the async return call goes straight to ReturnBidResult. So sounds like any semantic strategy that returns ShouldWait will cause resource-based selection to be skipped.

Reply:
ohh yeah, great catch! I've correct the code in the latest commits and added tests for this.